core/src/graph/actions.ts

```
@@ -1035,25 +1031,11 @@ const dependenciesFromActionConfig = profile(function dependenciesFromActionConf
   // -> We avoid depending on action execution when referencing static output keys
   const staticOutputKeys = definition?.staticOutputsSchema ? describeSchema(definition.staticOutputsSchema).keys : []
 
-  for (const ref of getActionTemplateReferences(config)) {
+  for (const ref of getActionTemplateReferences(config, templateContext)) {
     let needsExecuted = false
 
     const outputKey = ref.fullRef[4] as string
 
-    if (maybeTemplateString(ref.name)) {
-      try {
-        ref.name = resolveTemplateString({
-          string: ref.name,
-          context: templateContext,
-          contextOpts: { allowPartial: false },
-        })
-      } catch (err) {
-        log.warn(
```

Todo double check: did we retain old behaviour where we only warn?

---

I find it extremely unlikely that throwing instead of printing this confusing warning is a breaking change. is there a use case?

---

I suspect a couple of users may run into this in the context of action templates. Is there a particular reason to skip this entirely, instead of changing to match the new behavior?

---

I've managed to reproduce triggering the warning in `garden validate` using this template:

```
kind: Build
name: foobar2
type: exec
spec:
  command: [echo, '${actions.build[var.foo].outputs.log}']
variables:
  hello: {}
```

`garden build` fails with this error message (before my changes):
```
Invalid template string (${actions.build[var.foo].outputs.log}) at path command.1: Could not find key [object Object] under actions.build. Available keys: foobar2.
```

With the new changes the error now looks like this, and we fail already during validation:
```
Found invalid action reference: /Users/steffen/repro/project.garden.yml:22
...
20  | spec:
21  |   command: [echo, '${actions.build[var.foo].outputs.log}']
------------------------^
Could not find key foo under var. Available keys: hello.
```

The new error message is way more helpful compared to the warning we used to print (`⚠ graph                     → Unable to infer dependency from action reference in exec Build foobar2, because template string '${var.foo}' could not be resolved. Either fix the dependency or specify it explicitly.`. Failing earlier in the validation avoids the inevitable follow-up error when the action is being resolved.

I'll continue to look if I can find any issues with `ConfigTemplate`. If you have a scenario or example where the new approach will fail, please let me know